src/assets/scripts/client/aircraft/AircraftModel.js

```diff
@@ -1862,32 +1862,102 @@ export default class AircraftModel {
+        const nextAltitudeMaximumWaypoint = this.fms.findNextWaypointWithMaximumAltitudeRestriction();
+        const nextAltitudeMinimumWaypoint = this.fms.findNextWaypointWithMinimumAltitudeRestriction();
         const maximumAltitudeExists = !_isNil(nextAltitudeMaximumWaypoint);
         const minimumAltitudeExists = !_isNil(nextAltitudeMinimumWaypoint);
 
-        if (!maximumAltitudeExists && !minimumAltitudeExists) {
-            return this.mcp.altitude;
-        }
+        if (this.mcp.altitude < this.altitude) {
+            // we want to descend...
+            if (!minimumAltitudeExists || nextAltitudeMinimumWaypoint.altitudeMinimum <= this.mcp.altitude) {
+                // ... and there is nothing that can stop us.
+                return this.mcp.altitude;
+            }
 
-        if (maximumAltitudeExists && minimumAltitudeExists) {
-            const waypoints = this.fms.waypoints;
-            const indexOfMax = _findIndex(waypoints, (waypoint) => waypoint.name === nextAltitudeMaximumWaypoint.name);
-            const indexOfMin = _findIndex(waypoints, (waypoint) => waypoint.name === nextAltitudeMinimumWaypoint.name);
+            if (this.altitude < nextAltitudeMinimumWaypoint.altitudeMinimum) {
+                // ... but we are too low and we have to comply with VNAV restriction
+                return this._calculateTargetedAltitudeVnavClimb(nextAltitudeMinimumWaypoint);
+            }
 
-            if (indexOfMax < indexOfMin) {
+            if (maximumAltitudeExists) {
+                if (this.mcp.altitude > nextAltitudeMaximumWaypoint.altitudeMaximum) {
+                    // we are too high but we are prioritizing clearance over VNAV restriction
+                    return this.mcp.altitude;
+                }
+
+                if (this.altitude > nextAltitudeMaximumWaypoint.altitudeMaximum) {
+                    // we are too high...
+
+                    if (nextAltitudeMinimumWaypoint.altitudeMinimum > nextAltitudeMaximumWaypoint.altitudeMaximum) {
+                        // the minimum altitude is above the maximum altiude, check if we can descend all the way down
+                        // without violating VNAV restrictions.
+                        const firstWaypoint = this._findFirstWaypoint(this.fms.waypoints, nextAltitudeMinimumWaypoint, nextAltitudeMaximumWaypoint);
+
+                        if (firstWaypoint.name === nextAltitudeMinimumWaypoint.name) {
+                            // ... but we can not descend all the way down yet
+                            return nextAltitudeMinimumWaypoint.altitudeMinimum;
+                        }
+                    }
+
+                    // ...so descend to comply with VNAV restriction
+                    return this._calculateTargetedAltitudeVnavDescent(nextAltitudeMaximumWaypoint);
+                }
+            }
+        }
+        else {
```

Comment:
`else` should be on the same line as the closing `}`

Reply:
done